lib/auth/gitserver/gitserverv1/github.go

```diff
@@ -0,0 +1,133 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package gitserverv1
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/uuid"
+	"github.com/gravitational/trace"
+
+	apidefaults "github.com/gravitational/teleport/api/defaults"
+	pb "github.com/gravitational/teleport/api/gen/proto/go/teleport/gitserver/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/auth/integration/credentials"
+	"github.com/gravitational/teleport/lib/authz"
+)
+
+func (s *Service) CreateGitHubAuthRequest(ctx context.Context, in *pb.CreateGitHubAuthRequestRequest) (*types.GithubAuthRequest, error) {
+	if in.Request == nil {
+		return nil, trace.BadParameter("missing github auth request")
+	}
+	if err := types.ValidateGitHubOrganizationName(in.Organization); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	authCtx, gitServer, err := s.authAndFindServerByOrg(ctx, in.Organization)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	s.cfg.Log.DebugContext(ctx, "Creating GitHub auth request for authenticated user.",
+		"user", authCtx.User.GetName(),
+		"org", gitServer.GetGitHub().Organization,
+		"integration", gitServer.GetGitHub().Integration,
+	)
+
+	// Make a "temporary" connector spec and save it in the request.
+	uuid := uuid.NewString()
+	spec, err := s.makeGithubConnectorSpec(ctx, uuid, gitServer.GetGitHub().Organization, gitServer.GetGitHub().Integration)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	in.Request.ConnectorID = uuid
+	in.Request.ConnectorSpec = spec
+	in.Request.AuthenticatedUser = authCtx.User.GetName()
+	in.Request.CertTTL = authCtx.Identity.GetIdentity().Expires.Sub(s.cfg.clock.Now())
+	in.Request.ClientLoginIP = authCtx.Identity.GetIdentity().LoginIP
+
+	// More params of in.Request will get updated and checked by
+	// s.cfg.GitHubAuthRequestCreator.
+	request, err := s.cfg.GitHubAuthRequestCreator(ctx, *in.Request)
```

Comment:
> 	// More params of in.Request will get updated and checked by
	// s.cfg.GitHubAuthRequestCreator.

The in Request has many other fields that are validate by the old flow like; 
```
	if req.CreateWebSession && !a.hasBuiltinRole(types.RoleProxy) {
```

that are checked by the legacy https://github.com/gravitational/teleport/blob/de8e27f6a3d7f74e5ba4b176cba5b96ef9794507/lib/auth/auth_with_roles.go#L3952



By omitting these top level validations in new GRPC , the new flow introduces a potential security gap where the `CreateGitHubAuthRequest` requests could exploit the exposed method to bypass restrictions the are requried in CreateGithubAuthRequest. 


So I think we should double check the corner cases like bypassing CreateWebSession.

Reply:
Added some extra checks at top.